Add support for manila in multinode

Alex-Welsh opened the pull request:
Support Manila in the ci-multinode environment using the CephFS native backend. Configuration is automatic based on the Kolla variables to enable manila and the backend.

Alex-Welsh committed:
Add support for manila in multinode

Supports Manila in the ci-multinode environment using the CephFS native
backend. Configuration is automatic based on the Kolla variables to
enable manila and the backend.

Also moves cephadm-commands to run after other tasks so that the
commands can reference objects created earlier on.

Alex-Welsh commented:
Currently not fully tested

markgoddard commented on etc/kayobe/environments/ci-multinode/cephadm.yml:
```diff
@@ -35,9 +36,17 @@ cephadm_pools:
+cephadm_pools_manila_cephfs_native:
+  - name: cephfs_data
+    application: cephfs
+    state: present
```
Could use the state to make this conditional.

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/environments/ci-multinode/cephadm.yml:
```diff
@@ -56,3 +65,15 @@ cephadm_keys:
+    caps:
+      mon: "allow r"
+      mgr: "allow rw"
+    state: present
```
ditto

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/environments/ci-multinode/cephadm.yml:
```diff
@@ -35,9 +36,17 @@ cephadm_pools:
 
 # List of Cephx keys. See stackhpc.cephadm.keys role for format.
-cephadm_keys:
+cephadm_keys: "{{ cephadm_keys_default + (cephadm_keys_manila_cephfs_native if (kolla_enable_manila and kolla_enable_manila_backend_cephfs_native) else [] ) }}"
```
Generally best to apply the `bool` filter to any booleans to allow for `no`

Alex-Welsh replied:
I've done so on the grouping of the bools `(a and b) | bool` but I've just realised you might have meant do so on each bool individually e.g. `(a | bool and b | bool)`. I can change it round if needed

markgoddard replied:
It needs to be applied to each boolean variable

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/seed/network-interfaces:
```diff
@@ -10,6 +10,10 @@ external_interface: "{{ vxlan_interfaces[0].device }}.{{ external_vlan }}"
 
+# Only add the storage interface if we are using manila-cephfs.
+storage_interface_template: "{{ vxlan_interfaces[0].device }}.{{ storage_vlan }}"
+storage_interface: "{{ storage_interface_template if (kolla_enable_manila and kolla_enable_manila_backend_cephfs_native) else omit }}"
```
This can be defined unconditionally. You will also need to add `storage` to `seed_extra_network_interfaces`, which could be done conditionally.

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/cephadm.yml:
```diff
@@ -1,8 +1,8 @@
 ---
 # Deploy Ceph via Cephadm. Create EC profiles, CRUSH rules, pools and keys.
 - import_playbook: cephadm-deploy.yml
-- import_playbook: cephadm-commands.yml
```
I think Scott put it first to support setting size and min_size before creating pools.

Alex-Welsh replied:
I asked him about it and it was just to get rid of a health warning, so should be okay to run at any time

markgoddard replied:
I do think it would be necessary to run first if setting the size and min size though

Alex-Welsh replied:
I've split it into `cephadm_commands_pre` and `cephadm_commands_post` so we can run both

markgoddard replied:
Thanks

Alex-Welsh committed:
Add documentation for multinode manila

Alex-Welsh committed:
Split cephadm_commands into pre and post

Split `cephadm_commands` into `cephadm_commands_pre` and
`cephadm_commands_post` commands. This allows the user
to run commands that must be executed before the rest of
the post-deployment configuration, as well as commands
that rely on resources created by the post-deployment config.

markgoddard commented on doc/source/contributor/environments/ci-multinode.rst:
```diff
@@ -53,3 +53,153 @@ Final steps
+======
+The Multinode environment supports Manila with the CephFS native backend, but it
+is not enabled by default. To enable it, set the following in
+``/etc/kayobe/environments/ci-multinode/kolla.yml``:
```
```suggestion
``etc/kayobe/environments/ci-multinode/kolla.yml``:
```

Alex-Welsh replied:
Fixed in docs branch

markgoddard commented on doc/source/contributor/environments/ci-multinode.rst:
```diff
@@ -53,3 +53,153 @@ Final steps
+
+      ls testdir
+
+If it shows the test file then the share is working correctly.
```
Nice example test procedure

markgoddard commented on etc/kayobe/ansible/cephadm-commands-post.yml:
```diff
@@ -9,3 +9,5 @@
     - import_role:
         name: stackhpc.cephadm.commands
+      vars:
+        cephadm_commands: "{{ cephadm_commands_post }}"
```
I think we get accidental backwards compatibility here because if `cephadm_commands` is still defined in kayobe-config it will override this task var. (this is a good thing)

Alex-Welsh replied:
I think most ceph commands should be idempotent but when I do a follow up, I'll add a default to the `cephadm_commands_post` list so anything defined in `cephadm_commands` only gets run once, and should have exactly the same behaviour as before

markgoddard replied:
They should be idempotent anyway as it should be possible to run the playbook multiple times.

markgoddard commented on etc/kayobe/environments/ci-multinode/cephadm.yml:
```diff
@@ -35,6 +35,12 @@ cephadm_pools:
   - name: vms
     application: rbd
     state: present
+  - name: cephfs_data
+    application: cephfs
+    state: "{{ 'present' if (kolla_enable_manila | bool and kolla_enable_manila_backend_cephfs_native | bool) else 'absent' }}"
+  - name: cephfs_metadata
+    application: cephfs
+    state: "{{ 'present' if (kolla_enable_manila | bool and kolla_enable_manila_backend_cephfs_native | bool) else 'absent' }}"
```
Would be nice to add these to the s-k-c ceph docs after the existing example, saying "if using manila, you will need to add pools for it, e.g. ..."

This could be a follow up change

Alex-Welsh replied:
Done in docs branch

markgoddard commented on etc/kayobe/environments/ci-multinode/cephadm.yml:
```diff
@@ -56,3 +62,16 @@ cephadm_keys:
       osd: "profile rbd pool=images"
       mgr: "profile rbd pool=images"
     state: present
+  - name: client.manila
```
ditto above

Alex-Welsh replied:
Done in docs branch

markgoddard reviewed (approved):
Docs bits could be a follow up
